### PR TITLE
Use default values in POSIX-style variable expansions

### PR DESCRIPTION
**Pull request by gergelypolonkai**
An attempt to continue #30 with default values

It’s not perfect, but it works as i expect it to work.  For example, using `k` in `_replacement()` feels a bit hacky.

What do you think?

If you like the basic idea, i can add tests, documentation, etc. to this PR.

### CODE REVIEW

File: dotenv/main.py [hunk 1]
```diff
@@ -181,21 +181,42 @@ def unset_key(dotenv_path, key_to_unset, quote_mode="always"):
...
-        ret = os.getenv(name, new_values.get(name, ""))
-        return ret
+
+        # If name is set as an environment variable, use its value
+        if name in os.environ:
+            return os.getenv(name)
+
+        # If name is the same as k, we need to handle the situation special, or we will have ${K}
+        # or ${K:-default} as the return value.  Note that k comes from the for loop below
+        if name == k:
```
Comment (jamesmyatt) [inline dotenv/main.py]: `k` isn't defined

File: dotenv/main.py [hunk 2]
```diff
@@ -181,21 +181,42 @@ def unset_key(dotenv_path, key_to_unset, quote_mode="always"):
...
+            # If we have a default set, return it
+            if default:
```
Comment (jamesmyatt) [inline dotenv/main.py]: How about `return default or ""`?

File: dotenv/main.py [hunk 3]
```diff
@@ -181,21 +181,42 @@ def unset_key(dotenv_path, key_to_unset, quote_mode="always"):
 
 
 def resolve_nested_variables(values):
-    def _replacement(name):
+    def _replacement(name, default=None):
```
Comment (jamesmyatt) [inline dotenv/main.py]: How about some unit tests?

File: dotenv/main.py (hunk 1 above)
Comment (gergelypolonkai) [reply dotenv/main.py]: It is; it comes from the for loop that iterates over `values` right after the definition of `_re_sub_callback()`. Also see the end of the comment above this line.

That’s exactly the reason i wrote it feels hacky.

File: dotenv/main.py (hunk 2 above)
Comment (gergelypolonkai) [reply dotenv/main.py]: That was the original line of code; i don’t know why i changed it, but i will put that back in place.

File: dotenv/main.py (hunk 3 above)
Comment (gergelypolonkai) [reply dotenv/main.py]: If you like the general idea, i can go on and write the tests and expand the documentation with examples and caveats.

**Comment by gergelypolonkai** [comment]
It seems the tests fail on 3.3 regardless of my change. I wonder why it says anything about IPython…

File: dotenv/main.py (hunk 3 above)
Comment (gergelypolonkai) [reply dotenv/main.py]: If you like the general idea, i can go on and write tests and expand the documentation with the new functionality and some caveats.

Comment (jamesmyatt) [reply dotenv/main.py]: tests would help to clarify what the general idea is.

Note that I'm not a contributor to this library, so my opinion doesn't carry any weight.

File: dotenv/main.py [hunk 4]
```diff
@@ -181,21 +181,42 @@ def unset_key(dotenv_path, key_to_unset, quote_mode="always"):
...
+            # If we have a default set, return it
+            if default:
+                return default
+
+            # Otherwise, return an empty string
+            return ''
+
+        # If all the above failed, try getting name from the environments, revert to the value
+        # from values, and if it’s not in values yet, revert to the default value
+        return os.getenv(name, values.get(name, default or ''))
```
Comment (jamesmyatt) [inline dotenv/main.py]: It's already tried to get name from os.environ. Why try again here?

**Comment by bbc2** [comment]
Sorry I never really looked into this PR.  It is now superseded by #248.

**Pull request closed without merging**